apps/aecore/lib/aecore/channel/worker.ex

```diff
@@ -0,0 +1,526 @@
+  alias Aecore.Channel.ChannelStateOnChain
+  alias Aecore.Channel.ChannelStatePeer
+
+  alias Aecore.Channel.Tx.{
+    ChannelCloseMutalTx,
+    ChannelCloseSoloTx,
+    ChannelSlashTx,
+    ChannelSettleTx,
+    ChannelCreateTx
+  }
+
+  alias Aecore.Tx.{DataTx, SignedTx}
+  alias Aecore.Tx.Pool.Worker, as: Pool
+  alias Aecore.Keys.Wallet
+  alias Aeutil.Events
+
+  use GenServer
+
+  require Logger
+
+  @type role :: :initiator | :responder
+
+  @typedoc """
+  State is map channel_id -> channel_peer_state
+  """
+  @type state :: %{binary() => ChannelStatePeer.t()}
+
+  @type error :: {:error, binary()}
+
+  def start_link(_args) do
+    GenServer.start_link(__MODULE__, %{}, name: __MODULE__)
+  end
+
+  def init(_args) do
+    Events.subscribe(:new_top_block)
+    {:ok, %{}}
+  end
+
+  @doc """
+  Notifies channel manager about new mined tx
+  """
+  def new_tx_mined(%SignedTx{data: %DataTx{type: ChannelCreateTx}} = tx) do
+    opened(tx)
+  end
+
+  def new_tx_mined(%SignedTx{data: %DataTx{type: ChannelCloseMutalTx}} = tx) do
+    closed(tx)
+  end
+
+  def new_tx_mined(%SignedTx{data: %DataTx{type: ChannelCloseSoloTx}} = _tx) do
+    # TODO
+    :ok
+  end
+
+  def new_tx_mined(%SignedTx{data: %DataTx{type: ChannelSlashTx}} = _tx) do
+    # TODO
+    :ok
+  end
+
+  def new_tx_mined(%SignedTx{data: %DataTx{type: ChannelSettleTx}} = tx) do
+    closed(tx)
+  end
+
+  def new_tx_mined(_) do
+    # We don't care about this tx
+    :ok
+  end
+
+  @doc """
+  Imports channels from ChannelStatePeer object. Useful for storage
+  """
+  @spec import_channel(ChannelStatePeer.t()) :: :ok | error()
+  def import_channel(channel_state) do
+    id = ChannelStatePeer.id(channel_state)
+    GenServer.call(__MODULE__, {:import_channel, id, channel_state})
+  end
+
+  @doc """
+  Import channel from open tx. Assumes no transactions were made
+  """
+  @spec import_from_open(SignedTx.t(), non_neg_integer(), role()) :: :ok | error()
+  def import_from_open(open_tx, reserve, role) do
+    peer_state = ChannelStatePeer.from_open(open_tx, reserve, role)
+    import_channel(peer_state)
+  end
+
+  @doc """
+  Imports channel from open tx and ChannelStateOffChain.
+  """
+  @spec import_from_open_and_state(
+          SignedTx.t(),
+          ChannelStateOffChain.t(),
+          non_neg_integer(),
+          role()
+        ) :: :ok | error()
+  def import_from_open_and_state(open_tx, state, reserve, role) do
+    peer_state = ChannelStatePeer.from_open_and_state(open_tx, state, reserve, role)
+    import_channel(peer_state)
+  end
+
+  @doc """
+  Initializes channel with temporary ID. This has to be called for every channel by both :initiator and :responder.
+  """
+  @spec initialize(
+          binary(),
+          list(Wallet.pubkey()),
+          list(non_neg_integer()),
+          role(),
+          non_neg_integer()
+        ) :: :ok | error()
+  def initialize(temporary_id, pubkeys, amounts, role, channel_reserve) do
```

Comment:
Just as a suggestion for `pubkeys`.
Typically, when using the `list()` structure, this uniquely means that it is a "infinite" number of elements, and their ordering does not always matter. When it comes to a "certain/finite" number of items, `tuple()` are used, where the ordinance always matters.
Personally, I would use it like {initiator_key, responder_key}, if of course there are no obstacles.

But I'm not implying that the suggested solution is not right.

Reply:
Good idea. I will change it. (I had plans for making universal channels, but those were not incorporated by epoch and current solution that tries to be compatible with epoch has to handle both parties separately)